src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/CompiledPageMetadataProvider.cs

```diff
@@ -0,0 +1,82 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Mvc.ApplicationParts;
+using Microsoft.AspNetCore.Mvc.RazorPages.ApplicationFeature;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.Mvc.RazorPages.Internal
+{
+    public class CompiledPageMetadataProvider : IPageMetadataProvider
+    {
+        private readonly ApplicationPartManager _applicationManager;
+        private readonly RazorPagesOptions _pagesOptions;
+        private List<PageMetadata> _compiledPageMetadata;
+
+        public CompiledPageMetadataProvider(
+            ApplicationPartManager applicationManager,
+            IOptions<RazorPagesOptions> pagesOptionsAccessor)
+        {
+            _applicationManager = applicationManager;
+            _pagesOptions = pagesOptionsAccessor.Value;
+        }
+
+        public IEnumerable<PageMetadata> EnumeratePageMetadata()
+        {
+            EnsureCompiledPageMetadata();
+
+            return _compiledPageMetadata;
+        }
+
+        private void EnsureCompiledPageMetadata()
+        {
+            if (_compiledPageMetadata == null)
+            {
+                var feature = new CompiledPageInfoFeature();
+                _applicationManager.PopulateFeature(feature);
+
+                _compiledPageMetadata = new List<PageMetadata>(feature.CompiledPages.Count);
+
+                var rootDirectory = _pagesOptions.RootDirectory;
```

Comment:
I'm not sure why this would be relevant. It would be relevant when you're doing the view compilation.